File: 2022/day5.py
#!/usr/bin/python3

import sys

TEST = False
DAY = "5"

def tprint(msg):
    if TEST:
        print(msg)

def parse_crates(lines):
    stacks = {}
    stack_nums = []
    for stack_num in lines[-1].split():
        #print(stack_num)
        stacks[stack_num] = []
        stack_nums.append(stack_num)
    for line in lines[-2::-1]:
        i = 0
        for sn in stack_nums:
            crate = line[i:i+3]
            if crate[0] == "[" and crate[2] == "]":
                stacks[sn].append(crate[1])
            i += 4
    return stacks

def parse_moves(lines):
    moves = []
    for line in lines:
        parts = line.split()
        moves.append((int(parts[1]),parts[3],parts[5]))
    return moves

def parse_crates_and_moves(lines):
    crate_lines = []
    move_lines = []
    getting_crates = True
    for line in lines.splitlines():
        if line.strip() == "":
            getting_crates = False
            continue
        if getting_crates:
            crate_lines.append(line)
        else:
            move_lines.append(line)
    crates = parse_crates(crate_lines)
    moves = parse_moves(move_lines)
    return crates, moves

def repr_crates(crates):
    s = " ".join(" {} ".format(x) for x in sorted(crates.keys()))
    for i in range(max([len(v) for v in crates.values()])):
        these_crates = []
        for k in sorted(crates.keys()):
            if len(crates[k]) > i:
                these_crates.append("[" + crates[k][i] + "]")
            else:
                these_crates.append("   ")
        s = " ".join(these_crates) + "\n" + s
    return s


def part1(inp):
    crates, moves = parse_crates_and_moves(inp)
    #print(crates)
    #print(moves)
    for m in moves:
        for x in range(m[0]):
            tprint("{} -> {}".format(m[1],m[2]))
            crates[m[2]].append(crates[m[1]].pop())
    tprint(repr_crates(crates))
    tops = []
    for k in sorted(crates.keys()):
        tops.append(crates[k][-1])
    print("".join(tops))

def part2(inp):
    crates, moves = parse_crates_and_moves(inp)
    #print(crates)
    #print(moves)
    for m in moves:
        add_crates = []
        for x in range(m[0]):
            add_crates.append(crates[m[1]].pop())
        add_crates.reverse()
        for a in add_crates:
            crates[m[2]].append(a)
    tprint(repr_crates(crates))
    tops = []
    for k in sorted(crates.keys()):
        tops.append(crates[k][-1])
    print("".join(tops))

def main():
    global TEST
    filename = "inputs/day{}.input".format(DAY)
    if len(sys.argv) > 1 and (sys.argv[1] == "-t" or sys.argv[1] == "--test"):
        TEST = True
        filename = "inputs/day{}.test.input".format(DAY)
    inp = None
    with open(filename,'r') as fp:
        inp = fp.read()
    print("== Part 1 ==")
    part1(inp)
    print("== Part 2 ==")
    part2(inp)

if __name__ == "__main__":
    main()

File: 2022/day5.rs
use std::fs;

const DAY:&str = "5";

fn main() {
    let filename = format!("day{}.test.input",DAY);
    let contents = fs::read_to_string(filename)
        .expect("Expected to read file");
    println!("== Part 1 ==");
    part1(&contents);
    println!("== Part 2 ==");
    part2(&contents);
}

fn part1(contents: &str) {
    let mut stacks:Vec<Vec<u32>> = Vec::new();
    for i in 0..
    let mut stack_n : u8 = 0;
    let mut i : u32 = 0;
    let mut crates_i : u32 = 0;
    let mut buf : String = String::new();
    let mut read_state : u8 = 0;
    for c in contents.chars() {
        if read_state == 0 || read_state == 1 {
            buf.push_str(&c.to_string());
            crates_i += 1;
            if crates_i == 3 {
                println!("check buf for crate");
                println!("buf: {}",buf);
                if buf == "   " {
                    println!("buf is empty space");
                } else if buf.chars().nth(0).unwrap() == '[' && buf.chars().nth(2).unwrap() == ']' {
                    println!("buf is crate {}",buf.chars().nth(1).unwrap());

                } else if buf.chars().nth(1).unwrap() != ' ' {
                    println!("buf is crate index {}",buf.chars().nth(1).unwrap());
                    read_state = 1;
                }
                stack_n += 1;
            } else if crates_i == 4 {
                if c == '\n' {
                    stack_n = 0;
                    if read_state == 1 {
                        read_state = 2;
                    }
                }
                buf = String::new();
                crates_i = 0;
            }
        }
        i += 1;
    }
}

fn part2(contents: &str) {

}
